sharktank/sharktank/layers/paged_attention.py

```diff
@@ -367,64 +227,614 @@ def write_timestep(
+                block_seq_stride=block_seq_stride,
+                cache_dtype=cache_dtype,
+                device=device,
+            )
+            caches.append(cache)
+
+        self.caches = caches
+        self.cache_partition_count = cache_partition_count
+        self.devices = devices
+        self.attn_head_count = attn_head_count
+        self.attn_head_dim = attn_head_dim
+        self.block_seq_stride = block_seq_stride
+        self.transformer_block_count = transformer_block_count
+        self.shard_count = shard_count
+
+        self.unsharded_page_dims = [
+            self.transformer_block_count,
+            self.cache_partition_count,
+            self.block_seq_stride,
+            self.attn_head_count,
+            self.attn_head_dim,
+        ]
+
+        if self.devices is None:
+            self.devices = list(range(shard_count))
+
+    @property
+    def state_count(self):
+        return 1
+
+    def allocate(
+        self, page_count: int, devices: List[int] | None = None
+    ) -> List[SplitPrimitiveTensor]:
+        assert devices is None
+        shards = [cache.allocate(page_count)[0] for cache in self.caches]
+        return [SplitPrimitiveTensor(ts=shards, shard_dim=1, devices=self.devices)]
+
+    def shard_state(self, state: List[torch.Tensor]) -> List[SplitPrimitiveTensor]:
+        assert len(state) == 1
+        page_table = state[0].unflatten(1, self.unsharded_page_dims)
+
+        shards = []
+        head_start = 0
+        for cache in self.caches:
+            head_end = head_start + cache.attn_head_count
+            shard = page_table[:, :, :, :, head_start:head_end]
+            shard = shard.flatten(1)
+            shards.append(shard)
+            head_start = head_end
+
+        return [SplitPrimitiveTensor(ts=shards, shard_dim=1, devices=self.devices)]
+
+    def unshard_state(self, state: List[SplitPrimitiveTensor]) -> List[torch.Tensor]:
+        assert len(state) == 1
+        assert state[0].shard_count == len(self.caches)
+
+        state = [
+            cache.unshard_state([shard])[0]
+            for cache, shard in zip(self.caches, state[0].shards)
+        ]
+        state = SplitPrimitiveTensor(ts=state, shard_dim=4, devices=self.devices)
+
+        return [ops.unshard(state)]
+
+    def read(
+        self,
+        state: List[SplitPrimitiveTensor],
+        *,
+        transformer_block_index: int,
+        page_ids: ReplicatedTensor,
+    ):
+        assert len(state) == 1
+        assert state[0].shard_count == self.shard_count
+
+        for device in state[0].devices:
+            assert device in page_ids.devices
+
+        page_id_map = {d: i for i, d in enumerate(page_ids.devices)}
+        page_id_shards = [page_ids.shards[page_id_map[d]] for d in state[0].devices]
+
+        shards = []
+        for shard_state, cache, shard_page_ids in zip(
+            state[0].shards, self.caches, page_id_shards
+        ):
+            read = cache.read(
```

Comment:
`KVCache.read` returns `(key, value)`. so shards will be `[(k,v)_0, (k,v)_1, ...]`. What is like 349 below doing when it's indexing into each (k, v) based on `i in range(self.cache_partition_count)`?

Reply:
Discussed offline.